Added support for xlsm files

ShiroYasha18 opened the pull request:
## Description

1. Adding ```.xlsm``` to the supported formats in datamodel/base_models.py
2. Mapping ```.xlsm``` files to the``` MsExcelDocumentBackend``` in ```document_converter.py```
3. Adding a test file ```test_backend_msexcel_xlsm.py ```to verify the functionality
The current implementation treats .xlsm files similarly to``` .xlsx ```files without extracting VBA macros. Docling's ingestion is tightly bound to the ```DoclingDocument``` template  in  ``` docling-core``` repo for the output format, which doesn't currently have a way to store VBA macro information field.

This PR provides a temporary workaround by mapping ```.xlsm``` files to the same backend as .xlsx files, allowing them to be processed without the VBA features. This enables basic support for ```.xlsm``` files while a more comprehensive solution for handling VBA macros can be discussed with the core Docling repo maintainers!


This is in reference to a issue raised on data prep kit https://github.com/data-prep-kit/data-prep-kit/issues/1220

PeterStaar-IBM commented:
@ShiroYasha18 Hi, thank you for adding support to xlsm! I dont really understand why you have changes in the `docling/backend/docling_parse_v4_backend.py` file. Where are those changes used in xlsm?

ShiroYasha18 commented:
@PeterStaar-IBM  Thanks for replying. My apologies I was working on the issue #1581 at my end to test lru caching to speed things a little bit more . This has nothing to do with xlsm support. I think it got committed by me to this pr by mistake. I am not sure if you can cherry pick the xlsm support commits or should I correct it in the original fork branch and commit again

ShiroYasha18 commented:
Updated the fork branch I hope the changes to ```docling/backend/docling_parse_v4_backend.py``` got undo . I will test things around properly for issue #1581  then open another proper PR I think.

PeterStaar-IBM commented:
Excellent, I think now you need to update the `tests/test_backend_msexcel_xlsm.py`:

1. it needs to have a function starting with test
2. remove all `print` statements
3. add an explicit `assert {test}=={pred}`

ShiroYasha18 commented:
@PeterStaar-IBM hi why are these CI failing can u rerun these? I think I cannot re run these on my end

ShiroYasha18 commented:
@alexshmmy hiii, soo basically this fix was a workaround for the xlsm files as mentioned before it just maps these xlsm files to xlsx backend. Now coming to the macros, I am not really handling the macros for this pr because of the very tightly bound format of the paraquet in the docling-core repo. As per my knowledge the macros can be converted to txt from which it can be moved to a ```.docx``` format then we can convert it to .md I think using docling. Correct me if I am wrong but I dont think that .docx by default has any macros or VBA. Now unless we are talking about maybe xlsm or docm then yes we can extract vbas or macros to text then convert that to docx  and finally to .md format ig with docling.

alexshmmy commented:
@ShiroYasha18 Thank you for response. I can give an example. In 3gpp standardisation, there are documents with text in macros. For example this one:

https://www.3gpp.org/ftp/tsg_ran/WG1_RL1/TSGR1_120b/Docs//R1-2501739.zip 

you can see the words obeservations and proposals in page 3-4 are in macros. When covert with docling this fails. Any suggested approach to fix this?

cau-git commented on docling/datamodel/base_models.py:
```diff
@@ -34,11 +34,12 @@ class ConversionStatus(str, Enum):
 class InputFormat(str, Enum):
     """A document format supported by document backend parsers."""
 
+    PDF = "pdf"
```
Why is PDF added as an input format?

ShiroYasha18 replied:
I kind of deleted it by mistake soo added it back in the place

cau-git commented on docling/datamodel/base_models.py:
```diff
@@ -34,11 +34,12 @@ class ConversionStatus(str, Enum):
 
+    PDF = "pdf"
     DOCX = "docx"
+    XLSM = "xlsm"
```
Instead of treating `XLSM` as separate `InputFormat`, let's please remove this and add `xlsm` into `FormatToExtensions` mapping and a matching mimetype to `FormatToMimeType`.

ShiroYasha18 replied:
thanks for comment ill look into this

cau-git commented on docling/document_converter.py:
```diff
@@ -126,6 +126,9 @@ def _get_default_option(format: InputFormat) -> FormatOption:
         InputFormat.XLSX: FormatOption(
             pipeline_cls=SimplePipeline, backend=MsExcelDocumentBackend
         ),
+        InputFormat.XLSM: FormatOption(
```
This will not be needed with the changes suggested above.

cau-git commented on tests/test_backend_msexcel_xlsm.py:
```diff
@@ -0,0 +1,50 @@
+def convert_paths(obj):
+    if isinstance(obj, dict):
+        return {k: convert_paths(v) for k, v in obj.items()}
+    elif isinstance(obj, list):
+        return [convert_paths(i) for i in obj]
+    elif hasattr(obj, "__module__") and obj.__module__.startswith("pathlib"):
+        return str(obj)
+    else:
+        return obj
+
+def test_backend_msexcel_xlsm():
+
+    supported_ext = 'xlsm'
+    
+    input_files = [f for f in INPUT_DIR.iterdir() if f.is_file() and f.suffix[1:].lower() == supported_ext]
+
+    converter = DocumentConverter()
+    
+    xlsm_files_processed = 0
+    for file in input_files:
+        try:
+
+            result = converter.convert(str(file))
+     
+            assert result is not None, f"Conversion failed for {file.name}"
+            out_path = OUTPUT_DIR / (file.stem + ".json")
+            result_dict = convert_paths(result.model_dump())
+            with open(out_path, "w", encoding="utf-8") as f:
+                json.dump(result_dict, f, ensure_ascii=False, indent=2)
+            
+            assert out_path.exists(), f"Output file {out_path} was not created"
+            xlsm_files_processed += 1
+                
+        except Exception as e:
+            assert False, f"Failed to convert {file.name}: {e}"
+
+    if input_files:
+        assert xlsm_files_processed > 0, "No xlsm files were processed despite being present"
+    else:
+        assert True, "No xlsm files were found to process"
```
Please just update the existing [test_backend_msexcel](https://github.com/docling-project/docling/blob/main/tests/test_backend_msexcel.py) to include xlsm files, and add a tiny XSLM example file to the existing test dir.

cau-git reviewed (commented):
@ShiroYasha18 Thanks for this contribution! Can you please address the review comments below.

PeterStaar-IBM commented:
@ShiroYasha18 Could you run the styling,

```
poetry run pre-commit run --all-files
```

I believe we are nearly there to merge!

ShiroYasha18 commented:
@cau-git hello I have done your mentioned changes ! 
1. create a mime instead a full blown mapping to xlsx
2. Added a sample file for xlsm
3. Added comments guiding for xlsm file processing in ```test_backend_msexcel``` instead of writing another test file . There were few conflicts regarding size etc so left guiding comments for same .

Let me know if any other changes required :))
